Sources/MongoSwift/BSON/Document+Sequence.swift

```diff
@@ -0,0 +1,121 @@
+import Foundation
+import libmongoc
+
+/// An extension of `Document` to make it conform to the `Sequence` protocol.
+/// This allows you to iterate through the (key, value) pairs, for example:
+/// ```
+/// let doc: Document = ["a": 1, "b": 2]
+/// for (key, value) in doc {
+///     ...
+/// }
+/// ```
+extension Document: Sequence {
+    /// Returns a `DocumentIterator` over the values in this `Document`. 
+    public func makeIterator() -> DocumentIterator {
+        guard let iter = DocumentIterator(forDocument: self) else {
+            preconditionFailure("Failed to initialize an iterator over document \(self)")
+        }
+        return iter
+    }
+}
+
+/// An iterator over the values in a `Document`. 
+public class DocumentIterator: IteratorProtocol {
+    /// the libbson iterator. it must be a `var` because we use it as
+    /// an inout argument
+    internal var iter: bson_iter_t
+
+    /// Initializes a new iterator over the contents of `doc`. `doc` must remain alive for 
+    /// the lifetime of the iterator. Returns `nil` if the key is not found, or if an iterator 
+    /// cannot be created over `doc` due to an error from e.g. corrupt data.
+    internal init?(forDocument doc: Document) {
+        self.iter = bson_iter_t()
+        if !bson_iter_init(&self.iter, doc.data) { return nil }
+    }
+
+    /// Initializes a new iterator over the contents of `doc`. `doc` must remain alive for the
+    /// lifetime of the iterator. Returns `nil` if an iterator cannot be created over `doc` due
+    /// to an error from e.g. corrupt data, or if the key is not found.
+    internal init?(forDocument doc: Document, advancedTo key: String) {
+        self.iter = bson_iter_t()
+        if !bson_iter_init_find(&iter, doc.data, key.cString(using: .utf8)) {
+            return nil
+        }
+    }
+
+    /// Advances the iterator forward one value. Returns false if there is an error moving forward
+    /// or if at the end of the document. Returns true otherwise.
+    private func advance() -> Bool {
+        return bson_iter_next(&self.iter)
+    }
+
+    /// Returns the current key. Assumes the iterator is in a valid position.
+    internal var currentKey: String {
+        return String(cString: bson_iter_key(&self.iter))
+    }
+
+    /// Returns the current value. Assumes the iterator is in a valid position.
+    internal var currentValue: BsonValue? {
+        do {
+            switch self.currentType {
+            case .symbol:
+                return try Symbol.asString(from: self)
+            case .dbPointer:
+                return try DBPointer.asDocument(from: self)
```

Comment:
Ah, is this why the DBPointer and Symbol initializers can be internal? If so, aren't those initializers still throwing each time?